sw/device/lib/dif/register.h

```diff
@@ -0,0 +1,151 @@
+ *
+ * @param address an address to an MMIO register.
+ * @return a |reg32_t| value representing that register.
+ */
+inline reg32_t reg32_from_addr(uintptr_t address) {
+  return (reg32_t){
+      .inner_ptr = (volatile uint32_t *)address,
+  };
+}
+
+/**
+ * Reads an aligned word from the MMIO register |base| at the given byte offset.
+ *
+ * This function is guaranteed to commit a read to memory, and will not be
+ * reordered with respect to other register manipulations.
+ *
+ * @param base the register to read from.
+ * @param offset the offset to read at, in bytes.
+ * @return the read word.
+ */
+inline uint32_t reg32_read(reg32_t base, ptrdiff_t offset) {
+  return base.inner_ptr[offset / sizeof(uint32_t)];
+}
+
+/**
+ * Writes an aligned word to the MMIO register |base| at the given byte offset.
+ *
+ * This function is guaranteed to commit a write to memory, and will not be
+ * reordered with respect to other register manipulations.
+ *
+ * @param base the register to write to.
+ * @param offset the offset to write at, in bytes.
+ * @param value the word to write.
+ */
+inline void reg32_write(reg32_t base, ptrdiff_t offset, uint32_t value) {
+  base.inner_ptr[offset / sizeof(uint32_t)] = value;
+}
+
+/**
+ * Clears the bits in |mask| from the MMIO register |base| at the given offset.
+ *
+ * This function has the same guarantees as |reg32_read()| and |reg32_write()|.
+ *
+ * @param base the register to mask.
+ * @param offset the offset to apply the mask at, in bytes.
+ * @param mask the mask to clear from the selected register.
+ */
+inline void reg32_clear_mask(reg32_t base, ptrdiff_t offset, uint32_t mask) {
+  uint32_t value = reg32_read(base, offset);
+  value &= ~mask;
+  reg32_write(base, offset, value);
+}
+
+/**
+ * Sets the bits in |mask| from the MMIO register |base| at the given offset.
+ *
+ * This function has the same guarantees as |reg32_read()| and |reg32_write()|.
+ *
+ * @param base the register to mask.
+ * @param offset the offset to apply the mask at, in bytes.
+ * @param mask the mask to set on the selected register.
+ */
+inline void reg32_set_mask(reg32_t base, ptrdiff_t offset, uint32_t mask) {
+  uint32_t value = reg32_read(base, offset);
+  value |= mask;
+  reg32_write(base, offset, value);
+}
+
+/**
+ * Reads the bits in |mask| from the MMIO register |base| at the given offset.
+ * The difference from other register mask functions is that mask is not
+ * expected to be offset to the right position in the register by the caller.
+ *
+ * This function has the same guarantees as |reg32_read()| and |reg32_write()|.
+ *
+ * @param base the register to mask.
+ * @param offset the offset to apply the mask at, in bytes.
+ * @param mask the mask to read from the selected register.
+ * @param mask_offset the offset of the mask in the selected register.
+ * @retun return the value of the read mask.
+ */
+inline uint32_t reg32_read_mask(reg32_t base, ptrdiff_t offset, uint32_t mask,
```

Comment:
This function can be done more efficient - like reg32_get_bit. There is no need for shifting the value twice. Thanks @lenary